File: benchmarks/pandas/bench_regression.py
"""Benchmark: linregress + OLS regression on 10,000-row dataset.
Uses pure numpy to implement linregress and OLS (lstsq), matching tsb.
"""
import json
import time
import numpy as np

N = 10_000
WARMUP = 5
ITERATIONS = 20

x = np.arange(N) / N
noise = np.sin(np.arange(N) * 0.37) * 0.1
y = 2.5 * x + 1.2 + noise

# OLS setup: two predictors + intercept column
x2 = np.cos(np.arange(N) * 0.1)
X = np.column_stack([x, x2, np.ones(N)])


def numpy_linregress(x_arr, y_arr):
    """Pure-numpy simple linear regression (mirrors scipy.stats.linregress)."""
    n = len(x_arr)
    xm = x_arr.mean()
    ym = y_arr.mean()
    sxx = ((x_arr - xm) ** 2).sum()
    sxy = ((x_arr - xm) * (y_arr - ym)).sum()
    slope = sxy / sxx
    intercept = ym - slope * xm
    return slope, intercept


def ols_fit(X_mat, y_vec):
    """Numpy lstsq-based OLS (mirrors tsb OLS with addIntercept=True)."""
    coeffs, _, _, _ = np.linalg.lstsq(X_mat, y_vec, rcond=None)
    return coeffs


for _ in range(WARMUP):
    numpy_linregress(x, y)
    ols_fit(X, y)

start = time.perf_counter()
for _ in range(ITERATIONS):
    numpy_linregress(x, y)
    ols_fit(X, y)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "regression",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_regression.ts
/**
 * Benchmark: linregress + OLS regression on 10,000-row dataset.
 * Tests simple linear regression (linregress) and multiple OLS (.fit).
 */
import { linregress, OLS, Series, DataFrame } from "../../src/index.js";

const N = 10_000;
const WARMUP = 5;
const ITERATIONS = 20;

// Generate synthetic data
const x = Float64Array.from({ length: N }, (_, i) => i / N);
const noise = Float64Array.from({ length: N }, (_, i) => Math.sin(i * 0.37) * 0.1);
const y = Float64Array.from({ length: N }, (_, i) => 2.5 * x[i] + 1.2 + noise[i]);

const xArr = Array.from(x);
const yArr = Array.from(y);

// OLS setup: two predictors
const x2 = Float64Array.from({ length: N }, (_, i) => Math.cos(i * 0.1));
const X = DataFrame.fromArrays({ x1: xArr, x2: Array.from(x2) });
const ySeries = new Series(yArr);
const ols = new OLS();

// Warm up
for (let i = 0; i < WARMUP; i++) {
  linregress(xArr, yArr);
  ols.fit(X, ySeries);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  linregress(xArr, yArr);
  ols.fit(X, ySeries);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "regression",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
